### ISSUE TITLE
ARM64 support?

### ISSUE THREAD

**Issue by omh1280** [confluentinc/common-docker#117]
Is there a plan for supporting ARM with these images? I see [this issue](https://github.com/confluentinc/cp-docker-images/issues/718) and [this MR](https://github.com/confluentinc/cp-docker-images/pull/830) in the old repo, but they haven't been active in a while.

**Comment by omh1280** [comment]
@andrewegel are you the right person to ask about this? Are there any tasks you'd need help with to get this supported?

**Comment by andrewegel** [comment]
Taking a PR to enable ARM build support at this time could have the unintended side effect of meaning that Confluent On Prem Docker Images supports ARM, but that (currently) would not be true from the business' perspective. I'm currently not aware of any sort of ARM support on Confluent's road map, but you're likely not the first or last to ask this question. I apologize that I can't definitively give you an answer in certain terms when (if) ARM support is coming, but the ARM architecture movement in the industry hasn't gone unnoticed by Confluent.

**Comment by tzachshabtay** [comment]
@andrewegel can't there be a comment in the doc stating that ARM support is experimental and/or not official? that should be enough for now, I would think.

**Comment by andrewegel** [comment]
> @andrewegel can't there be a comment in the doc stating that ARM support is experimental and/or not official? that should be enough for now, I would think.

"I can't definitively give you an answer in certain terms when (if) ARM support is coming"

**Comment by ferencbeutel4711** [comment]
ARM support is also important for the osx m1 users with the new macbooks. Is there anything planned to support these devices or a workaround specific to the m1 macbooks?

**Comment by lmyslinski** [comment]
If anyone else is looking for a workaround, @jonstelly has prepared some arm64 kafka images that I found to be working nicely. I've replaced 'confluentinc/cp-kafka' with 'kymeric/cp-kafka' for my testcontainers tests and everything passed. More details here: https://github.com/confluentinc/cp-docker-images/pull/830

**Comment by etolbakov** [comment]
@jonstelly Jon, thank you for creating `kymeric/cp-base`.
If anyone needs an Arm64 `schema-registry` image that works on Arm64 I'd like to share my workaround. 
Either you can build the image on your laptop replacing schema-registry `Dockerfile` in [confluentinc/cp-docker-images](https://github.com/confluentinc/cp-docker-images/blob/5.3.1-post/debian/schema-registry/Dockerfile) with 
https://gist.github.com/etolbakov/92c77f0c4cd4602afc7ce3800d5e1c65
or feel free to go with `eugenetea/schema-registry-arm64:latest` straightaway.

**Comment by rakibulalam** [comment]
It almost wasted more than 15 days of my working effort.  Is there any plan for the arm64?

**Comment by alexcu** [comment]
It does appear that installing _plugins_ also fails on arm64.

I am attempting to install a Debezium plugin on the `cp-kafka-connect-base` image. The Debezium Postgres plugin will _download_ correctly via `confluent-hub`, but will not _load_ correctly:

```
[2021-07-15 02:50:39,770] INFO Loading plugin from: /usr/share/confluent-hub-components/debezium-debezium-connector-postgresql (org.apache.kafka.connect.runtime.isolation.DelegatingClassLoader:246)
[2021-07-15 02:50:40,377] ERROR Stopping due to error (org.apache.kafka.connect.cli.ConnectDistributed:86)
java.lang.ExceptionInInitializerError
	at io.debezium.connector.postgresql.PostgresConnector.version(PostgresConnector.java:45)
	at org.apache.kafka.connect.runtime.isolation.DelegatingClassLoader.versionFor(DelegatingClassLoader.java:387)
	at org.apache.kafka.connect.runtime.isolation.DelegatingClassLoader.versionFor(DelegatingClassLoader.java:392)
	at org.apache.kafka.connect.runtime.isolation.DelegatingClassLoader.getPluginDesc(DelegatingClassLoader.java:362)
	at org.apache.kafka.connect.runtime.isolation.DelegatingClassLoader.scanPluginPath(DelegatingClassLoader.java:334)
	at org.apache.kafka.connect.runtime.isolation.DelegatingClassLoader.scanUrlsAndAddPlugins(DelegatingClassLoader.java:268)
	at org.apache.kafka.connect.runtime.isolation.DelegatingClassLoader.registerPlugin(DelegatingClassLoader.java:260)
	at org.apache.kafka.connect.runtime.isolation.DelegatingClassLoader.initPluginLoader(DelegatingClassLoader.java:229)
	at org.apache.kafka.connect.runtime.isolation.DelegatingClassLoader.initLoaders(DelegatingClassLoader.java:206)
	at org.apache.kafka.connect.runtime.isolation.Plugins.<init>(Plugins.java:61)
	at org.apache.kafka.connect.cli.ConnectDistributed.startConnect(ConnectDistributed.java:93)
	at org.apache.kafka.connect.cli.ConnectDistributed.main(ConnectDistributed.java:80)
Caused by: java.lang.NullPointerException: inStream parameter is null
	at java.base/java.util.Objects.requireNonNull(Objects.java:246)
	at java.base/java.util.Properties.load(Properties.java:406)
	at io.debezium.util.IoUtil.loadProperties(IoUtil.java:493)
	at io.debezium.util.IoUtil.loadProperties(IoUtil.java:523)
	at io.debezium.connector.postgresql.Module.<clinit>(Module.java:19)
	... 12 more
```

If I _manually_ download the plugin's tar.gz and extract to `/usr/share/confluent-hub-components`, then the plugin will load correctly. However, then I get this different issue (`Could not initialize class org.eclipse.jetty.http.MimeTypes$Type`) when admin resources are added, and the **connector does not start**:

```
[2021-07-17 02:19:19,538] INFO Adding admin resources to main listener (org.apache.kafka.connect.runtime.rest.RestServer:241)
[2021-07-17 02:19:20,009] ERROR Stopping due to error (org.apache.kafka.connect.cli.ConnectDistributed:86)
java.lang.NoClassDefFoundError: Could not initialize class org.eclipse.jetty.http.MimeTypes$Type
	at org.eclipse.jetty.http.MimeTypes.<clinit>(MimeTypes.java:175)
	at org.eclipse.jetty.server.handler.ContextHandler.doStart(ContextHandler.java:895)
	at org.eclipse.jetty.servlet.ServletContextHandler.doStart(ServletContextHandler.java:288)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:73)
	at org.apache.kafka.connect.runtime.rest.RestServer.initializeResources(RestServer.java:305)
	at org.apache.kafka.connect.runtime.Connect.start(Connect.java:55)
	at org.apache.kafka.connect.cli.ConnectDistributed.startConnect(ConnectDistributed.java:140)
	at org.apache.kafka.connect.cli.ConnectDistributed.main(ConnectDistributed.java:80)
[2021-07-17 02:19:20,020] INFO Kafka Connect stopping (org.apache.kafka.connect.runtime.Connect:67)
```

As a sanity check, I switched to the Postgres plugin to the MySQL plugin, it will download and load correctly (using either `confluent-hub`), but the error above persists. Attempting to switch to [Confluent's own JDBC connector](https://www.confluent.io/hub/confluentinc/kafka-connect-jdbc) plugin results in the same behaviour as well, whether it be installed using `confluent-hub` or manually downloaded and extracted.

There is a bit of back-and-forth in this [community hub post](https://forum.confluent.io/t/debezium-kafka-connect-error-loading-plugin-from-confluent-hub-via-docker/2217/14).

Would be good if anyone can replicate on their M1 machine, and see if similar behaviour is observable. 

**TL;DR:**

Plugin | Install Type | docker-compose.yml | logs | plugin loads? | connector starts?
---|---|---|---|---|---
Debezium PostgreSQL | confluent-hub | [link](https://gist.github.com/alexcu/9c46a5f946a025120ca2b89e67de5e2b#file-docker-compose-postgres-confluent-hub-yml) | [link](https://gist.github.com/alexcu/9c46a5f946a025120ca2b89e67de5e2b#file-connector-confluent-hub-log) | ❌  | ❌ 
Debezium PostgreSQL | manual | [link](https://gist.github.com/alexcu/9c46a5f946a025120ca2b89e67de5e2b#file-docker-compose-postgres-manual-yml) | [link](https://gist.github.com/alexcu/9c46a5f946a025120ca2b89e67de5e2b#file-connector-manual-log) | ✅ | ❌ 
Debezium MySQL | confluent-hub | [link](https://gist.github.com/alexcu/9c46a5f946a025120ca2b89e67de5e2b#file-docker-compose-mysql-confluent-hub-yml) | [link](https://gist.github.com/alexcu/9c46a5f946a025120ca2b89e67de5e2b#file-connect-mysql-confluent-hub-log)  | ✅ | ❌ 
Debezium MySQL | manual | [link](https://gist.github.com/alexcu/9c46a5f946a025120ca2b89e67de5e2b#file-docker-compose-mysql-manual-yml) | [link](https://gist.github.com/alexcu/9c46a5f946a025120ca2b89e67de5e2b#file-connect-mysql-manual-log) | ✅ | ❌ 
Confluent JDBC | confluent-hub | [link](https://gist.github.com/alexcu/9c46a5f946a025120ca2b89e67de5e2b#file-docker-compose-jdbc-confluent-hub-yml) | [link](https://gist.github.com/alexcu/9c46a5f946a025120ca2b89e67de5e2b#file-connect-jdbc-confluent-hub-log) | ✅ | ❌ 
Confluent JDBC | manual | [link](https://gist.github.com/alexcu/9c46a5f946a025120ca2b89e67de5e2b#file-docker-compose-jdbc-manual-yml) | [link](https://gist.github.com/alexcu/9c46a5f946a025120ca2b89e67de5e2b#file-connect-jdbc-manual-log) | ✅ | ❌ 

Open to any further ideas if anyone has any clues...

**Comment by SasonBraha** [comment]
Any update?

**Comment by aviflax** [comment]
Just in case it’s helpful: I’ve been building my own images locally (and using them in my own `docker-compose` file) and they’re working well for me on my M1 Mac. They’re not nearly as sophisticated or configurable as the images produced by this repo, and right now they’ll only launch the broker and schema registry successfully, but [these files (GitHub Gist)](https://gist.github.com/aviflax/25c1196177722b6a93169df685618316) at the very least demonstrate a pattern that others could follow.

**Comment by agates4** [comment]
https://github.com/localstack/localstack/issues/4456#issuecomment-905904064

I see an issue with alpine linux JDK on M1 with an illegal charset - that problem is afflicting confluent images

**Comment by georgelza** [comment]
... 2 wishes  here also... want to build/deploy the entire cp stack on my to be MAC, currently have a i7, but about to pull the trigger on the new M1... so need a ARM compatible set of docker images for the CP stack and then #2... I'm looking at putting together a little lab at home out of 4 or 5 RPi4's with K8S deployed, on which I want to deploy the full stack also.
Connector capability (Redis, ElasticSearch, PostgreSQL & MysSQL at min) and schema manager is important.
G

**Comment by fractalzombie** [comment]
It will be great to get support for M1, because for now it's working through amd64 and eats up a lot of memory...

**Comment by chr0non** [comment]
It will be great to get support for M1, because everyone is using M1 for work. It's very uncomfortable after 10 months of waiting to receive "segmentation fault" while other popular docker packages are already been ported to arm64...

**Comment by jecacs** [comment]
Omg, one year after release apple silicon, and it still not working

**Comment by georgelza** [comment]
... and with all the server platforms also now becoming ARM based... ARM is becoming enterprise, not just a at home desktop toy...

**Comment by agates4** [comment]
@andrewegel from your earlier message you stated the business currently doesn't have plans to widely announce ARM support - is this still the case?

**Comment by jecacs** [comment]
Hire me, and I make ARM support if you can't

**Comment by rohit2b** [comment]
Hi folks,

Confluent has released Docker images for `arm64` for development and testing purposes. Check it out across different Confluent Platform components in Docker Hub: https://hub.docker.com/u/confluentinc

For example, here's how to pull the latest CP Kafka image:

```docker pull confluentinc/cp-kafka:latest.arm64```
